[2.19] Add content-hashing to pub docs

MaryaBelanger opened the pull request:
Fixes #4361 

Changes:

- Added a best practices entry about `--enforce-lockfile` to [pub/dependencies](https://dart.dev/tools/pub/dependencies#best-practices) page (per @jonasfj suggestion [here](https://github.com/dart-lang/site-www/issues/4361#issuecomment-1331865826))
- Added `--enforce-lockfile` to [pub-get#options](https://dart.dev/tools/pub/cmd/pub-get#options) section
- Adjusted the [pub/glossary#lockfile](https://dart.dev/tools/pub/glossary#lockfile) section to discuss content hashing
- Adjusted the [pub/versioning#lockfiles](https://dart.dev/tools/pub/versioning#lockfiles) section to discuss content hashing (per @parlough suggestion [here](https://github.com/dart-lang/site-www/issues/4361#issuecomment-1314727160))

I'm fairly confident these are the right _places_ to add info on these new features, but am not confident in the accuracy/quality of the information I added. I think I might've conflated what content hashes actually do vs what the `--enforce-lockfile` flag does. 

Looking forward to reviews to get this right!

domesticmouse commented on src/tools/pub/dependencies.md:
```diff
@@ -441,21 +441,34 @@ Use [caret syntax](#caret-syntax)
-Test whenever you update package dependencies
-: If you run [`pub upgrade`][] without updating your pubspec,
+### Test whenever you update package dependencies
+ If you run [`pub upgrade`][] without updating your pubspec,
```
Should this be `dart pub upgrade` or `flutter pub upgrade`? The standalone `pub` command is no more...

parlough replied:
Yeah we tend to prefer just `dart pub ...` on the Dart site now, especially if linking to the command.

domesticmouse commented on src/tools/pub/dependencies.md:
```diff
@@ -441,21 +441,34 @@ Use [caret syntax](#caret-syntax)
 
+### Verify the integrity of downloaded packages
+ When getting new dependencies, use the [`--enforce-lockfile`][]
+  option to ensure the extracted package content matches
```
Is the indent level change intended?

MaryaBelanger replied:
Honestly I don't know 😅 

If you look at [the current md](https://github.com/dart-lang/site-www/blob/main/src/tools/pub/dependencies.md?plain=1#L434), there's some kind of colon-indent formatting going on. But if you look at the [rendered page](https://dart.dev/tools/pub/dependencies#best-practices) it doesn't seem to actually format  anything? So I kept the indents just in case but that's probably not necessary.

domesticmouse reviewed (approved):
Couple of minor questions

parlough commented:
We need a good place to link [`/go/content-hashes`](https://github.com/dart-lang/site-www/blob/main/firebase.json#L142) to. Without a page for `pubspec.lock` currently, perhaps a new "Content hashes" entry in the glossary? Then that could be linked to from the lockfile entry as well.

MaryaBelanger commented:
> We need a good place to link [`/go/content-hashes`](https://github.com/dart-lang/site-www/blob/main/firebase.json#L142) to. Without a page for `pubspec.lock` currently, perhaps a new "Content hashes" entry in the glossary? Then that could be linked to from the lockfile entry as well.

I was worried about the go/ link too. A "Content hashes" section in the glossary is a GREAT idea. It'll look a lot better to remove the awkward interjections I wrote into the lockfiles sections and link to a dedicated section on the new feature instead.

Like I mentioned in the PR description, I'm operating on low understanding of the topic currently so it'll probably be a bit sparse to begin with, but that's fine it can be filled out later. 

Thanks for the great idea @parlough!

MaryaBelanger commented:
@parlough I'm not sure how to check that the go/ redirect works, could you check?

parlough commented:
> @parlough I'm not sure how to check that the go/ redirect works, could you check?

You can just add the redirect to the end of the automatically staged site: https://dart-dev--pr4407-pub-content-hashing-b8vb4esy.web.app/go/content-hashes

Looks like it works, thanks for adding the glossary entry and adjusting the redirect to it :)

sigurdm commented on src/tools/pub/glossary.md:
```diff
@@ -29,6 +29,27 @@ specify `any` for their dependencies' [version constraints][].
 
+## Content hashes
+
+The first time `dart pub get` runs,
+it computes a sha256 hash for each hosted package,
+derived from the contents of their [lockfiles][].
+The hash embeds in the description of `pubspec.lock`.
+The pub client uses content hashes to
+verify the integrity of downloaded archives
+and detect if something has changed on the server.
+On future `pub get` runs, 
+the client can verify descriptions provided by the server
+against the descriptions of its cached packages.
+If they don't match, the archive is redownlaoded.
+If they still don't match, the resolution fails with an error.
+
+To utilize this feature during package resolution,
+call [`dart pub get --enforce-lockfile`][].
+
+[lockfiles]: #lockfile
+[`dart pub get --enforce-lockfile`]: /tools/pub/cmd/pub-get#--enforce-lockfile
```
```suggestion
The pub.dev repository serves the sha256 hash of each package version it hosts.
This is useful for the pub client to validate the integrity of the downloaded package. 

When `dart pub get` downloads a package, it computes the hash of the downloaded archive and re-downloads if there is a discrepancy. 

But furthermore the hash is used to protect against changes on the repository.

The hash of each hosted dependency will be stored with the resolution in the [lockfile][].

The pub client uses this content hash to verify that exactly the same packages are used when running `pub get` again using the same lock file, potentially on a different computer.

If the locked hash doesn't match what's currently in the pub cache, the archive is redownloaded.
If it still doesn't match, the lockfile is updated and a warning is printed like:
```
$ dart pub get
Resolving dependencies...
Cached version of foo-1.0.0 has wrong hash - redownloading.
 ~ foo 1.0.0 (was 1.0.0)
The existing content-hash from pubspec.lock doesn't match contents for:
 * foo-1.0.0 from "pub.dev"
This indicates one of:
 * The content has changed on the server since you created the pubspec.lock.
 * The pubspec.lock has been corrupted.
 
The content-hashes in pubspec.lock has been updated.

For more information see:
https://dart.dev/go/content-hashes

Changed 1 dependency!
```

This updated hash will now show up in your version control diff and should make you suspicious.

To make a discrepancy become an error instead of a warning use [`dart pub get --enforce-lockfile`][] that will fail if it cannot find package archives with the same hashes without updating the lockfile.

```
$ dart pub get --enforce-lockfile
Resolving dependencies...
Cached version of foo-1.0.0 has wrong hash - redownloading.
~ foo 1.0.0 (was 1.0.0)
The existing content-hash from pubspec.lock doesn't match contents for:
 * foo-1.0.0 from "pub.dev"

This indicates one of:
 * The content has changed on the server since you created the pubspec.lock.
 * The pubspec.lock has been corrupted.

For more information see:
https://dart.dev/go/content-hashes
Would change 1 dependency.
Unable to satisfy `pubspec.yaml` using `pubspec.lock`.

To update `pubspec.lock` run `dart pub get` without
`--enforce-lockfile`.
```

[lockfile]: #lockfile
[`dart pub get --enforce-lockfile`]: /tools/pub/cmd/pub-get#--enforce-lockfile
```

MaryaBelanger replied:
This is excellent @sigurdm thanks so much for writing it up!

sigurdm reviewed (commented):
I made a rewrite of the content-hashes section. Maybe it is getting slightly too long - but I think the details are important to get right here.

@jonasfj surely also has opinions here.

atsansone commented on src/tools/pub/cmd/pub-get.md:
```diff
@@ -145,6 +145,13 @@ By default, pub precompiles executables
 
+### `--enforce-lockfile`
+
+Fetches dependencies, but will refuse to resolve
+if anything deviates from `pubspec.lock` or there is no `pubspec.lock`.
+This is useful for deploying,
+and ensuring reproducible runs in CI and production. 
```
**issue:** Tighten this language to ease readability and match later revised copy.

See your copy in [this file](https://github.com/dart-lang/site-www/pull/4407/files#diff-f24f8dca36386ceb9bc6e11d49921934f1486e12084bed60f621cd772348cd41R464-R471).

**question:** What does "resolve" mean here and in the linked copy? What is being resolved?

MaryaBelanger replied:
Added the part about is causing an error as opposed to just a warning. 

> question: What does "resolve" mean here and in the linked copy? What is being resolved?

I was going to add a link to an explanation of package resolution elsewhere in the docs, but I realized the link I would've added is _to this page_. `pub get` itself _is the definition_ for package resolution; everything it does is what package resolution...is. Linking to the page we're already on seemed like overkill. We can assume that someone reading about `--enforce-lockfile` knows what "resolve" means in this context.

atsansone commented on src/tools/pub/dependencies.md:
```diff
@@ -429,33 +429,50 @@ depend on that package.
+
+### Use [caret syntax](#caret-syntax)
+
+Specifying dependencies with version ranges is such as `^1.6.3`
+is a good practice because it allows the pub tool to
+select newer versions of the package when they become available.
```
**issues:** Revise this first sentence. Issues includes:

- A misplaced word repetition
- A length of more than 30 words
- A redundancy (This whole section is about best practices; no need to restate that.)
- Missing backticks for a command name

atsansone commented on src/tools/pub/dependencies.md:
```diff
@@ -429,33 +429,50 @@ depend on that package.
+Specifying dependencies with version ranges is such as `^1.6.3`
+is a good practice because it allows the pub tool to
+select newer versions of the package when they become available.
+Also, it places an upper limit on the allowed version,
+based on an assumption that packages use [semantic versions][],
+where any version of path versioned `1.x` is compatible,
+but where a new version `2.x` would be a major upgrade
+that isn't semantically compatible with `1.x` versions. 
```
**issues:** Revise this sentence. Issues include:
- Length of almost 50 words
- One idea that needs its dependent clause close, but maybe in a separate sentence
- Another dependent clause that could be one or two other sentences

atsansone commented on src/tools/pub/dependencies.md:
```diff
@@ -429,33 +429,50 @@ depend on that package.
+
+### Depend on the latest stable package versions
+
+Use [`dart pub upgrade`][] to update to the latest package versions
+that your pubspec allows.
```
**suggestion:** Flip the sentence.
```suggestion
To update your app or package to the latest versions of their dependencies,
use [`dart pub upgrade`][].  The `pubspec` file lists these dependencies. 
```

atsansone commented on src/tools/pub/dependencies.md:
```diff
@@ -429,33 +429,50 @@ depend on that package.
+
+Use [`dart pub upgrade`][] to update to the latest package versions
+that your pubspec allows.
+To identify dependencies in your app or package that
+aren't on the latest stable versions,
+use [`dart pub outdated`][].
```
**suggestion:** Tighten this language. There should be a more concise way to write this. Maybe some copy could be added prior to this that makes clear that any time you say "app", you could mean "app or package" so you don't need to repeat it too much.

atsansone commented on src/tools/pub/dependencies.md:
```diff
@@ -429,33 +429,50 @@ depend on that package.
+
+### Test whenever you update package dependencies
+
+If you run [`dart pub upgrade`][] without updating your pubspec,
+the API should stay the same
+and your code should run as before—but test to make sure.
```
**issue** The way this is written sounds like you are not encouraging testing. If you are, start with that, then say that the API and code should be unaffected anyway.

atsansone commented on src/tools/pub/dependencies.md:
```diff
@@ -429,33 +429,50 @@ depend on that package.
+If you run [`dart pub upgrade`][] without updating your pubspec,
+the API should stay the same
+and your code should run as before—but test to make sure.
+If you modify the pubspec and update to a new major version,
+then you might encounter breaking changes,
+so you need to test even more thoroughly.
```
**issue:** This also reads as if you are backing into your point. Lead with the fact that _they should always test, *especially* if they modify the `pubspec` file._

atsansone commented on src/tools/pub/dependencies.md:
```diff
@@ -429,33 +429,50 @@ depend on that package.
+
+### Verify the integrity of downloaded packages
+
+When getting new dependencies, use the [`--enforce-lockfile`][]
+option to ensure the extracted package content matches
+the contents of the original archive.
+Without modifying the [lockfile][],
+this flag only resolves new dependencies if:
+* `pubspec.yaml` is satisfied
+* `pubspec.lock` is not missing
+* The packages' [content hashes][] match
```
See [this comment](https://github.com/dart-lang/site-www/pull/4407/files#diff-ab1f08ea6448a1b6e767a02de133569ca302eea3fa3807931043da0e661da631R153).

MaryaBelanger replied:
That comment also links back to this comment. I'm confused, is the issue with "resolve"? It's the normal definition of "resolve" applied to packages - you have a package already, and you're getting new info on it, so the system resolves the difference between new and existing to decide what's needed. This isn't the place to define package resolution

atsansone replied:
Both places cover almost the same ground, but do so in different ways. This second copy block explains it _better_, so I'm unsure why they differ.

atsansone commented on src/tools/pub/glossary.md:
```diff
@@ -29,6 +29,74 @@ specify `any` for their dependencies' [version constraints][].
 
+## Content hashes
+
+The pub.dev repository serves a sha256 hash of each package version it hosts.
```
**question:** "Serves" reads odd here. How is a hash served? Is it listed with the package? Is it included as part of the download? Also, looking to the next sentence, make it clear that `pub` is doing this validation, that is, The `pub` client validates a package's integrity using this hash.

MaryaBelanger replied:
>  Also, looking to the next sentence, make it clear that pub is doing this validation, that is, The pub client validates a package's integrity using this hash.

The second sentence already says literally that exact same thing: 

> "The pub client can use this hash to validate the integrity of downloaded packages,
> and protect against changes on the repository." 

So I'm not sure what you're asking for there.

atsansone commented on src/tools/pub/glossary.md:
```diff
@@ -29,6 +29,74 @@ specify `any` for their dependencies' [version constraints][].
+The updated content hash will show up in your version control diff,
+and should make you suspicious.
+
+To make a discrepancy become an error instead of a warning, use
+[`dart pub get --enforce-lockfile`][]. It will cause the resolution to fail
+if it cannot find package archives with the same hashes, without updating the lockfile.
```
**issue:** Include some version of this in the explanation of `--enforce-lockfile`.

MaryaBelanger replied:
Which part exactly? It seems to me to say the exact same thing: "...will refuse to resolve
if anything deviates from `pubspec.lock`"

atsansone replied:
This copy block provides context to resolution: package hashes don't match and you can't update the lock file. This is a great explanation.

atsansone reviewed (changes requested):
@MaryaBelanger I had some questions and issues within this PR. I'm left unsure what happens in validating packages as written. You do have the complete picture, but it's scattered throughout this update.

MaryaBelanger commented:
@atsansone I didn't write the entirety of the "Best practices" section of `dependencies.md`.

The diff is only that I removed some odd indentation formatting, and added the "Verify the integrity..." section. So I'm going to resolve all the comments on the other sections for the sake of time and keeping this PR on topic. But feel free to open another issue or PR to track editing the rest of the page.

atsansone commented:
@MaryaBelanger : Fair. Sounds good.

MaryaBelanger commented on src/tools/pub/cmd/pub-get.md:
```diff
@@ -145,6 +145,13 @@ By default, pub precompiles executables
+Fetches dependencies, but will refuse to resolve
+if anything deviates from `pubspec.lock`.
+This is useful for deploying to prodcution,
+and ensuring reproducible runs in CI and production. 
```
Skipping:
- The engineer specifically removed "deploying _to production_", not technically correct
- I think it's important to keep the dependent clause as is, "Fetches dependencies BUT (does other thing that doesn't normally happen when fetching dependencies)" is the very important point of the tool. Breaking the clauses up muddles the definition

atsansone commented on src/tools/pub/cmd/pub-get.md:
```diff
@@ -145,6 +145,13 @@ By default, pub precompiles executables
+
+Fetches dependencies, but will refuse to resolve
+if anything deviates from `pubspec.lock` or there is no `pubspec.lock`.
+Such deviations normally trigger a warning, but
+`--enforce-lockfile` turns it into an error.
```
**suggestion:** Consider reversing the flow of this sentence to remove the qualifier.

```suggestion
Including `--enforce-lockfile` changes what would
be a warning into an error. This error prevents Dart
from resolving the dependencies until you remove
any deviation from the `pubspec.lock` file.
```

MaryaBelanger replied:
Removing deviations from pubspec.lock is not an aspect of this, not going to incoporate

atsansone reviewed (approved):
@MaryaBelanger : LGTM % 2 small issues with the first change. One is per the Style Guide, then other would be for clarity. Make these changes and all is good.

parlough commented on src/tools/pub/glossary.md:
```diff
@@ -29,6 +29,74 @@ specify `any` for their dependencies' [version constraints][].
+downloaded archive. The hash of each hosted dependency is stored with the
+[resolution][] in the [lockfile][].
+
+The pub client uses this content hash to verify that running `pub get` again using
+the same lockfile, potentially on a different computer, uses exactly the same packages.
```
Maybe "exactly the same package" could be made a bit more specific, since it could protect from matching packages, just modified or outdated.

I know pub uses the term ["package archive"](https://github.com/dart-lang/pub/blob/master/doc/repository-spec-v2.md#list-all-versions-of-a-package) to describe the saved contents of the file, but perhaps that would sound confusing here. Any thoughts?

I'll think about it some more, but if no one has ideas, I suppose we can just leave this as is.

MaryaBelanger replied:
Maybe "uses the same packages in the same state", or "uses the correct package states"?

I don't really understand the topic enough to give a better suggestion

parlough commented on src/tools/pub/glossary.md:
```diff
@@ -111,16 +179,20 @@ identifying information for every immediate and transitive dependency a package
 The lockfile is generated automatically for you by pub when you run
 [`pub get`](/tools/pub/cmd/pub-get), [`pub upgrade`](/tools/pub/cmd/pub-upgrade),
-or [`pub downgrade`](/tools/pub/cmd/pub-downgrade).
+or [`pub downgrade`](/tools/pub/cmd/pub-downgrade). Pub also computes a [content
+hash][] to check against future resolutions.
```
Perhaps this is less of an "also" and more of a first class functionality:

```suggestion
or [`pub downgrade`](/tools/pub/cmd/pub-downgrade). 
Pub includes a [content hash][] for each package
to check against during future resolutions.
```

parlough commented on src/tools/pub/versioning.md:
```diff
@@ -341,8 +341,8 @@ result is a complete list of every package that your app depends on either
 
-Pub takes that and writes it out to a **lockfile** in your app's directory
-called `pubspec.lock`. 
+Pub takes that information, computes a [content hash][] from it,
```
This reads like it computes just one content hash of the whole list of dependencies, but doesn't it do it for each package?

MaryaBelanger replied:
Correct, added "For each package..." to the beginning

parlough commented on src/tools/pub/versioning.md:
```diff
@@ -341,8 +341,8 @@ result is a complete list of every package that your app depends on either
-Pub takes that and writes it out to a **lockfile** in your app's directory
-called `pubspec.lock`. 
+Pub takes that information, computes a [content hash][] from it,
+and writes both to a [**lockfile**][] in your app's directory called `pubspec.lock`. 
```
I don't think this link will work since it doesn't match the definition later. You should be able to pull out the bold:

```suggestion
and writes both to a **[lockfile][]**
in your app's directory called `pubspec.lock`. 
```

MaryaBelanger replied:
Hmm, that didn't work but adding the asterisks to the definition at the bottom does.

parlough reviewed (commented):
Thanks for working on this! It really ended up need modifications in a lot of different areas. You really thought of everything, nice job.

I have a few suggestions, questions, and comments.

MaryaBelanger commented on src/tools/pub/dependencies.md:
```diff
@@ -429,33 +429,50 @@ depend on that package.
+select newer versions of the package when they become available.
+Also, it places an upper limit on the allowed version,
+based on an assumption that packages use [semantic versions][],
+where any version of path versioned `1.x` is compatible,
```
I don't know what this means. I'm not using this PR to edit the whole page, I just changed the formatting because it was relevant to the section I added. But that should be tracked with a different issue since both you and @atsansone brought up issues with the rest of the best practices.

MaryaBelanger commented on src/tools/pub/dependencies.md:
```diff
@@ -429,33 +429,50 @@ depend on that package.
+the contents of the original archive.
+Without modifying the [lockfile][],
+this flag only resolves new dependencies if:
+* `pubspec.yaml` is satisfied
```
That's how I interpreted it. [From the implementation](https://github.com/dart-lang/pub/pull/3482#:~:text=will%20refuse%20to%20resolve%20if%20pubspec.yaml%20isn%27t%20satisfied%2C)

MaryaBelanger commented on src/tools/pub/glossary.md:
````diff
@@ -29,6 +29,74 @@ specify `any` for their dependencies' [version constraints][].
+pub redownloads the archive. If it still doesn't match, the lockfile
+updates and a warning is printed. For example:
+
+```
````
Good idea!

parlough commented on src/tools/pub/glossary.md:
```diff
@@ -111,16 +181,21 @@ identifying information for every immediate and transitive dependency a package
 The lockfile is generated automatically for you by pub when you run
 [`pub get`](/tools/pub/cmd/pub-get), [`pub upgrade`](/tools/pub/cmd/pub-upgrade),
 or [`pub downgrade`](/tools/pub/cmd/pub-downgrade).
+Pub inlcudes a [content hash][] for each package
```
```suggestion
Pub includes a [content hash][] for each package
```
